File: pysrc/fractal/brain/_op/parameters.py
from .._brain import Brain


class Parameters(
    Brain,
):
    def parameters(self):
        """
        Return the collection of trainable parameters of a network.
        """
        raise NotImplementedError

    def regularized_parameters(self):
        """
        Return the collection of regularized parameters of a network.
        This usually excludes neuron's biases.
        """
        raise NotImplementedError

    def number_of_parameters(self):
        """
            num_parameters(::AbstractNetwork)

        Return the total number of parameters of a network.
        """
        return sum(len(p) for p in self.parameters())

    def number_of_regularized_parameters(self):
        """
        Return the total number of regularized parameters of a network.
        """
        return sum(len(p) for p in self.regularized_parameters())

File: src/fractal/brain/_op/parameters.rs
from .._brain import Brain

pub trait Plasticity {
    fn parameters(&self);
    //! Return the collection of trainable parameters of a brain.
    
    fn regularized_parameters(&self);
    //! Return the collection of regularized parameters of a network.
    //! This usually excludes neuron's biases.
    
    fn number_of_parameters(&self);
    //! Return the total number of parameters of a network.

    fn number_of_regularized_parameters(&self);
    //! Return the total number of regularized parameters of a network.

impl Plasticity for Brain {
    fn parameters(&self) {}

    fn regularized_parameters(&self) {}

    fn number_of_parameters(&self) {
        return sum(len(p) for p in &self.parameters())
    }

    fn number_of_regularized_parameters(&self) {
        return sum(len(p) for p in &self.regularized_parameters())
    }
}
